cpp/docs/DEVELOPER_GUIDE.md

```diff
@@ -745,6 +745,22 @@ void isolated_helper_function(...);
 
 [**Anonymous namespaces should *never* be used in a header file.**](https://wiki.sei.cmu.edu/confluence/display/cplusplus/DCL59-CPP.+Do+not+define+an+unnamed+namespace+in+a+header+file)
 
+# Deprecating and Removing Code
+
+To minimize disruption for users, any breaking changes must be signaled by deprecation warnings for one release prior to the change.
+This includes wholesale removals of functions/methods/classes *and* changes to existing function signatures.
+Deprecations should be indicated using the [deprecated](https://en.cppreference.com/w/cpp/language/attributes/deprecated) attribute.
+Deprecated APIs should be documented using Doxygen's [deprecated](https://www.doxygen.nl/manual/commands.html#cmddeprecated) command.
+
+When an API is deprecated, if a replacement exists the replacement should be mentioned in both the deprecation message and the deprecation documentation.
+Changes to function parameters should be accompanied by the introduction of suitable overloads to avoid API breaks for one release cycle.
+Where possible, deprecated APIs should call new APIs, but duplication in deprecated code is acceptable at the discretion of PR reviewers if avoiding it imposes an undue burden on developers.
+
+PRs containing deprecations should be labeled as "non-breaking".
```

Comment:
This is where the whole "breaking" vs. "non-breaking" thing starts to "break down" :) 

Using a `[[deprecated]]` keyword on an API will cause a compiler warning by default. If a user has warnings as errors enabled, their build is broken until they disable that warning or update to the new API. Was that a breaking change?

Reply:
I chose to ignore this issue in what I wrote up. Feel free to suggest an alternative if you have a suggestion on what might be better to write here. I think using `[[deprecated]]` is still better than not, since at least for people not compiling with `Wall` we'll be offering a gentler off ramp. For people who are, the situation is no different than an outright removal.